### ISSUE TITLE
Handle SAML logout requests sent by the IdP

### ISSUE THREAD

**Issue by leleuj** [pac4j/pac4j#876]
Certainly around the `SAML2DefaultResponseValidator`...

**Comment by leleuj** [comment]
@jkacer : any update on this?

**Comment by jkacer** [comment]
Hi! So far only a tiny merge request for the demo application - better logout URL construction: https://github.com/pac4j/j2e-pac4j-demo/pull/27 ; would you mind having a look?
I'm sorry to be so slow; I'm quite short of time. But definitely going to work on this.

**Comment by leleuj** [comment]
No problem. I just merged your changes. We'll release 2.0 very soon, so maybe it will be a 2.0.1 fixes version.

**Comment by jkacer** [comment]
I've just looked at it little deeper and it seems the error occurs already a few lines up the stack. Of course, the exception in the SAML2ResponseValidator is obvious, because the message to be processed is of type LogoutResponseImpl, which does not implement Response (both from OpenSaml).

But should the code really get to DefaultCallbackLogic? Shouldn't it get to another type of logic? Something like DefaultLogoutLogic but for finishing the logout. Because DefaultCallbackLogic always calls getCredentials() on the chosen client, that's probably not what we want.

The caller of DefaultCallbackLogic is CallbackFilter, as it can be seen from this stack trace:
```
at org.pac4j.saml.sso.impl.SAML2DefaultResponseValidator.validate(SAML2DefaultResponseValidator.java:128)
at org.pac4j.saml.sso.impl.SAML2WebSSOMessageReceiver.receiveMessage(SAML2WebSSOMessageReceiver.java:77)
at org.pac4j.saml.sso.impl.SAML2WebSSOProfileHandler.receive(SAML2WebSSOProfileHandler.java:35)
at org.pac4j.saml.client.SAML2Client.lambda$0(SAML2Client.java:110)
at org.pac4j.core.client.BaseClient.retrieveCredentials(BaseClient.java:61)
at org.pac4j.core.client.IndirectClient.getCredentials(IndirectClient.java:123)
at org.pac4j.core.engine.DefaultCallbackLogic.perform(DefaultCallbackLogic.java:79) [replace already this ???]
at org.pac4j.j2e.filter.CallbackFilter.internalFilter(CallbackFilter.java:62)

```
Shouldn't the filter somehow decide which callback to call? Right now it always calls DefaultCallbackLogic. 

Or do you think DefaultCallbackLogic should handle this type of event too? If it should, I'll have to adapt it.

Maybe you discussed it with Matthieu when he was contributing the SLO code and I just did not pay enough attention; I'm sorry if this is the case.

**Comment by taggiasco** [comment]
Hi !
In fact, the DefaultLogoutLogic appeared just after my PR. However I don't know if it should be used.
How would you determiner in CallbackFilter what callback to use ?
I think the goal of the DefaultLogoutLogic is to handle logout (from the LogoutFilter), when the user wants to log out.
I think the URI for the SAML Single Logout should be different from the URI for the login process. As they are defined in the metadata, we could have different ones. For the login process, we come to callback and thus DefaultCallbackLogic, and for the logout, we come to ...logout... and thus the DefaultLogoutLogic applies. And so, it's the job of the LogoutFilter, and thus the DefaultLogoutLogic to know what to do with the request.
Am I wrong ?

**Comment by jkacer** [comment]
Hi and thank you for your comment!

I think I understand it little more now :-) So the response from the IdP should probably end up in DefaultLogoutLogic, not DefaultCallbackLogic. As you pointed out, I will probably have to change the Logout URL in the metadata in order to to route it to DefaultLogoutLogic. And then, a new validator should be used for LogoutResponse.

Thanks! I'll keep you updated...

**Comment by leleuj** [comment]
I disagree on this:

- the `/logout` endpoint (`DefaultLogoutLogic`) is meant to trigger logout processes

- the `/callback` endpoint (`DefaultCallbackLogic`)  is meant to receive callbacks from the identity providers.

For CAS (which handles all logout features: SLO, back/front-channel logouts), the callback endpoint is used to receive the login callback, but also the back-channel and front-channel logouts. See this class: https://github.com/pac4j/pac4j/blob/master/pac4j-cas/src/main/java/org/pac4j/cas/credentials/extractor/TicketAndLogoutRequestExtractor.java#L59

I think we should stick to that.

**Comment by taggiasco** [comment]
Ok I see.

So in this case, the callback logic will handle login response, logout response (when logout initialized from SP) and logout request (when logout initialized from IdP).

That sounds good for me.

**Comment by jkacer** [comment]
OK, got it, thank you! So no change of the URL in the metadata, DefaultCallbackLogic will handle this.

**Comment by leleuj** [comment]
@jkacer Do you have any idea when we'll be good on this?

**Comment by jkacer** [comment]
Please do not wait for me with a release, if you can omit this... I can't make any prediction.

**Comment by leleuj** [comment]
No problem. We'll merge your changes when they'll be available (likely after the release).

**Comment by jkacer** [comment]
Hi! I have written some code, so far only locally, it's far from being complete/working, and I just would like to validate the concept with you.

It works like this:
- `DefaultCallbackLogic` now calls a new method `client.decideAction(context)` and based on the result it either gets the credentials as before, performs logout (logout finish - after IdP replies) or performs logout (reaction to IdP-initiated logout).
- The method `decideAction(context)` normally returns constant GET_CREDENTIALS for all clients, just SAML clients behave specifically and examine the context. They create a SAML context from the provided web context and get the SAML object from it. Based on the type of the object (Response, LogoutResponse, LogoutRequest), they return one of [GET_CREDENTIALS, FINISH_LOGOUT, INIT_LOGOUT_FROM_OTHER_SIDE] that DefaultCallbackLogic will decide upon.
- `Client` also has `verifyLogout(context)` that will examine the message and throw an exception if it is found incorrect. Only SAML clients do something here.
- Apart message verification, `DefaultCallbackLogic` will perform actions similar to local logout for the two logout actions: `manager.logout()` etc.

Do you think this concept is viable? Or do you find some obstacles / design issues here? Thank you!

**Comment by leleuj** [comment]
To be honest, I had in mind something simpler like the solution we already have for CAS which relies on the `TicketAndLogoutRequestExtractor`: https://github.com/pac4j/pac4j/blob/master/pac4j-cas/src/main/java/org/pac4j/cas/credentials/extractor/TicketAndLogoutRequestExtractor.java

That said, what exists is too limited and not well-designed.

What you propose looks better, it's a very interesting discussion. And we certainly need to go further to handle all the use cases.

Currently, we have:

1) The `DefaultCallbackLogic` gets the credentials and user profile (`getCredentials` and `getUserProfile` methods of the client).

2) The `DefaultLogoutLogic` handles the local logout using the `manager.logout()` call and triggers the central logout via the `getLogoutAction` method (depending on the configuration).

I'm not sure to clearly understand the difference between: "performs logout (logout finish - after IdP replies) or performs logout (reaction to IdP-initiated logout)". Can you elaborate on this?

For CAS, we currently handle back-channel or front-channel logout calls in the application from the CAS server.

I think we need the `CallbackAction decideCallbackAction(context)` and `HttpAction handleLogout(context)` methods in the client (the `CallbackAction` being an `enum`) , which delegates the work to the `CallbackActionEvaluator` and `LogoutHandler` components. You may find better names.

I fear this may lead to breaking changes, in that case, we'll wait for the version 3.0.0 but I keep the hope we can make it backward compatible.

Let's submit a pull request to get into the details.

**Comment by mmoayyed** [comment]
FYI I am working on a deployment that needs this. I might be able to lend a hand and pick this up next week.

**Comment by leleuj** [comment]
Great! Keep us posted

**Comment by jkacer** [comment]
I'll try to create a simple merge request (against my master, just to show you the changes) this week.

**Comment by mmoayyed** [comment]
Update: Work is put on hold until further notice.

**Comment by leleuj** [comment]
@jkacer @mmoayyed have we made any progress on this?

**Comment by jkacer** [comment]
Hi! No, unfortunately not, I haven't managed to put it into a consistent + working state and haven't worked on it since June. But we are implementing SAML Single Logout at IDC right now; the SP --> IdP direction seems to be fine now; we have a plan to do IdP --> SP as well and this task would be a part of it. Unfortunately no exact schedule yet, it depends on the management when I will be allowed to work on it.

**Comment by leleuj** [comment]
No problem. Don't hesitate to keep us posted (even if it is to say you haven't had any time). Keeping the issue open.

**Comment by Caltalys** [comment]
At line 103 in [SAML2LogoutResponseValidator](https://github.com/pac4j/pac4j/blob/e36ceb65b3a7cfb17abe0f922ff1074df9a5080b/pac4j-saml/src/main/java/org/pac4j/saml/sso/impl/SAML2LogoutResponseValidator.java#L103) i think it should be type of [org.opensaml.saml.saml2.core.LogoutResponse]()

**Comment by leleuj** [comment]
This might be. Have you tested this feature?

**Comment by krzysztof-zych** [comment]
Hello.
Any progress here? I am looking for this feature and I am ready to help at any time if you need to.

**Comment by leleuj** [comment]
I have received a lot of requests on the SAML missing features, I'm working on them, and if you are able to test them, I'll ping you when things are ready!

**Comment by krzysztof-zych** [comment]
Sure, I can test them or I can help you with the code.

**Comment by krzysztof-zych** [comment]
Hi @leleuj 
Sorry to bother you, I just want to ask do you have any information about progress and approximate time of finish this task?

**Comment by mmoayyed** [comment]
We recently added SAML2 SLO support to CAS as an IDP, and it would be good for pac4j to gain similar support here. I have done some brief experiments with SS-SAML, but I can't quite trust the tests I did since SS-SAML is based on opensaml-2 at this point. I am going to pick up the pac4j side of things for the change and see how I might make progress, but I have no estimate on a timeline. If I were to guess, I'd say sometime definitely in 2018.
